ipa-core/src/report/hybrid.rs

```diff
@@ -22,6 +83,114 @@ where
     breakdown_key: Replicated<BK>,
 }
 
+impl<BK: SharedValue> Serializable for HybridImpressionReport<BK>
+where
+    BK: SharedValue,
+    Replicated<BK>: Serializable,
+    <Replicated<BK> as Serializable>::Size: Add<U16>,
+    <<Replicated<BK> as Serializable>::Size as Add<<Replicated<BA64> as Serializable>::Size>>:: Output: ArrayLength,
+{
+    type Size = <<Replicated<BK> as Serializable>::Size as Add<<Replicated<BA64> as Serializable>::Size>>:: Output;
+    type DeserializationError = InvalidHybridReportError; //<Replicated<BK> as Serializable>::DeserializationError;
+
+    fn serialize(&self, buf: &mut GenericArray<u8, Self::Size>) {
+        let mk_sz = <Replicated<BA64> as Serializable>::Size::USIZE;
+        let bk_sz = <Replicated<BK> as Serializable>::Size::USIZE;
+
+        self.match_key
+            .serialize(GenericArray::from_mut_slice(&mut buf[..mk_sz]));
+
+        self.breakdown_key
+            .serialize(GenericArray::from_mut_slice(&mut buf[mk_sz..mk_sz + bk_sz]));
+    }
+    fn deserialize(buf: &GenericArray<u8, Self::Size>) -> Result<Self, Self::DeserializationError> {
+        let mk_sz = <Replicated<BA64> as Serializable>::Size::USIZE;
+        let bk_sz = <Replicated<BK> as Serializable>::Size::USIZE;
+        let match_key =
+            Replicated::<BA64>::deserialize(GenericArray::from_slice(&buf[..mk_sz]))
+            .map_err(|e| InvalidHybridReportError::DeserializationError("match_key", e.into()))?;
+        let breakdown_key =
+            Replicated::<BK>::deserialize(GenericArray::from_slice(&buf[mk_sz..mk_sz + bk_sz]))
+            .map_err(|e| InvalidHybridReportError::DeserializationError("breakdown_key", e.into()))?;
+        Ok(Self { match_key, breakdown_key })
+    }
+}
+
+impl<BK> HybridImpressionReport<BK>
+where
+    BK: SharedValue,
+    Replicated<BK>: Serializable,
+    <Replicated<BK> as Serializable>::Size: Add<U16>,
+    <<Replicated<BK> as Serializable>::Size as Add<<Replicated<BA64> as Serializable>::Size>>:: Output: ArrayLength,
+{
+    const BTT_END: usize = <Replicated<BK> as Serializable>::Size::USIZE;
+
+    /// # Panics
+    /// If report length does not fit in `u16`.
+    pub fn encrypted_len(&self) -> u16 {
+        let len = EncryptedHybridImpressionReport::<BK, &[u8]>::SITE_DOMAIN_OFFSET;
+        len.try_into().unwrap()
+    }
+    /// # Errors
+    /// If there is a problem encrypting the report.
+    pub fn encrypt<R: CryptoRng + RngCore>(
+        &self,
+        key_id: KeyIdentifier,
+        key_registry: &impl PublicKeyRegistry,
+        rng: &mut R,
+    ) -> Result<Vec<u8>, InvalidHybridReportError> {
+        let mut out = Vec::with_capacity(usize::from(self.encrypted_len()));
```

Comment:
how large are these reports? Does `out` fit in memory?

Reply:
I think these should be pretty small. In the test it's something like 116 bytes. I'd check with @eriktaubeneck though

Reply:
I thought `out` would contain all the reports. Not worried about the test use case, but more concerned with the real use case, when we have 1 billion events.

Reply:
Oh, I see. The implementation for OprfReport uses a Vec. Are the hybrid reports expected to be larger?

Reply:
HybridReports are likely a little smaller than OprfReports. This method only encrypts an individual report, so somewhere between 100-150 bytes. 

When we are generating a lot of them (for the 1B case), we will have to worry about memory (probably streaming them into a file, for example), but that happens elsewhere. Using a Vec for individual reports seems to be the right approach as we cannot know the actual length at compile time, due to some of the associated data being arbitrary length strings.

Reply:
I think this is a good design because we also have `encrypt_to` that allows to use reusable buffer for encryptions if this ever becomes a bottleneck